File: android/app/src/main/java/com/faceproject/android/ui/screens/profile/ProfileViewModel.kt
package com.faceproject.android.ui.screens.profile

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.faceproject.android.data.ApiResult
import com.faceproject.android.data.AuthRepository
import com.faceproject.android.model.HeartRateStatus
import com.faceproject.android.model.UserProfile
import kotlinx.coroutines.launch

data class HeartRateUiState(
    val available: Boolean = false,
    val bpm: Double? = null,
    val status: HeartRateStatus = HeartRateStatus.UNKNOWN,
    val isMeasuring: Boolean = false
)

sealed class ProfileUiState {
    data object Loading : ProfileUiState()
    data class Loaded(val profile: UserProfile) : ProfileUiState()
    data class Error(val message: String) : ProfileUiState()
}

class ProfileViewModel(
    private val repository: AuthRepository = AuthRepository()
) : ViewModel() {

    var uiState by mutableStateOf<ProfileUiState>(ProfileUiState.Loading)
        private set

    var heartRate by mutableStateOf(HeartRateUiState())
        private set

    fun load(name: String) {
        uiState = ProfileUiState.Loading

        viewModelScope.launch {
            when (val result = repository.getUser(name)) {
                is ApiResult.Success -> uiState = ProfileUiState.Loaded(result.data)
                is ApiResult.Error -> uiState = ProfileUiState.Error(result.message)
            }
        }

        viewModelScope.launch {
            when (val result = repository.latestHeartRate(name)) {
                is ApiResult.Success -> {
                    val bpm = result.data.bpm
                    heartRate = HeartRateUiState(
                        available = result.data.available,
                        bpm = bpm,
                        status = bpm?.let { HeartRateStatus.fromBpm(it) } ?: HeartRateStatus.UNKNOWN
                    )
                }
                // Heart-rate is best-effort; a failure here shouldn't block the profile screen.
                is ApiResult.Error -> heartRate = HeartRateUiState()
            }
        }
    }

    fun measure(name: String) {
        heartRate = heartRate.copy(isMeasuring = true)

        viewModelScope.launch {
            val result = repository.measureHeartRate(name)
            heartRate = when (result) {
                is ApiResult.Success -> {
                    val bpm = result.data.bpm
                    HeartRateUiState(
                        available = result.data.available,
                        bpm = bpm,
                        status = bpm?.let { HeartRateStatus.fromBpm(it) } ?: HeartRateStatus.UNKNOWN
                    )
                }
                is ApiResult.Error -> heartRate.copy(isMeasuring = false)
            }
        }
    }
}

File: ios/FaceAuth/Views/ProfileViewModel.swift
import Foundation

struct HeartRateUiState {
    var available: Bool = false
    var bpm: Double? = nil
    var status: HeartRateStatus = .unknown
    var source: String? = nil
    var isMeasuring: Bool = false
}

@MainActor
final class ProfileViewModel: ObservableObject {
    enum State {
        case loading
        case loaded(UserProfile)
        case error(String)
    }

    @Published var state: State = .loading
    @Published var heartRate = HeartRateUiState()

    private let healthKit = HealthKitManager()

    func load(name: String) {
        state = .loading

        Task {
            switch await APIClient.shared.getUser(name: name) {
            case .success(let profile):
                state = .loaded(profile)
            case .failure(let message):
                state = .error(message)
            }
        }

        Task {
            switch await APIClient.shared.latestHeartRate(name: name) {
            case .success(let response):
                heartRate = Self.mapHeartRate(response)
            case .failure:
                heartRate = HeartRateUiState()
            }
        }
    }

    func measureFromWatch(name: String) {
        heartRate.isMeasuring = true

        Task {
            let bpm: Double?
            let source: String?

            switch await healthKit.readLatestHeartRate() {
            case .success(let value, _):
                bpm = value
                source = "apple_watch"
            // HealthKit 미지원(예: 시뮬레이터)이거나 워치에서 동기화된 값이
            // 없는 경우 — 기존 "측정 준비 중" 응답으로 자연스럽게 대체된다.
            case .noData, .notAvailable, .failure:
                bpm = nil
                source = nil
            }

            switch await APIClient.shared.measureHeartRate(name: name, bpm: bpm, source: source) {
            case .success(let response):
                heartRate = Self.mapHeartRate(response)
            case .failure:
                heartRate.isMeasuring = false
            }
        }
    }

    private static func mapHeartRate(_ response: HeartRateResponse) -> HeartRateUiState {
        let status = response.bpm.map(HeartRateStatus.from(bpm:)) ?? .unknown
        return HeartRateUiState(
            available: response.available,
            bpm: response.bpm,
            status: status,
            source: response.source,
            isMeasuring: false
        )
    }
}
